File: Day287/countNodesEqualToAverageOfSubtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int count  =0;
    private int[] find(TreeNode root) {
        if (root == null) {
            return new int[]{0,0};
        }
        int left[] = find(root.left);
        int right[] = find(root.right);
        int sum = left[0]+right[0] +root.val;
        int c = left[1] + right[1] + 1;
        if (c!=0 && sum/c == root.val) {
            count++;
        }
        return new int[]{sum, c};
    }
    public int averageOfSubtree(TreeNode root) {
        find(root);
        return count;
    }
}

File: Day287/countNodesEqualToAverageOfSubtree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func inorder(root *TreeNode, sum *int, count *int) {
    if root == nil {
        return
    }
    inorder(root.Left, sum,count)
    *sum += root.Val
    *count++
    inorder(root.Right, sum,count)
}

func averageOfSubtree(root *TreeNode) int {
    queue := make([]*TreeNode, 0)
    answer := 0
    queue = append(queue, root)
    for len(queue) > 0 {
        // for _, v := range queue {
        //     fmt.Println(v.Val)
        // }
        sum := 0
        count := 0
        temp := queue[0]
        queue  = queue[1:]
        inorder(temp, &sum, &count)
        // fmt.Println(sum)
        if temp.Val == sum/count {
            answer++
        }
        if temp.Left != nil {
            queue = append(queue, temp.Left)
        }
        if temp.Right != nil {
            queue = append(queue, temp.Right)
        }
    }
    return answer
}